File: backend/handlers/compatibility.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"
)

type partDetails struct {
	Details map[string]interface{} `json:"details"`
}

type compatibilityRequest struct {
	Parts map[string]partDetails `json:"parts"`
}

type compatibilityResponse struct {
	Issues []string `json:"issues"`
}

// CompatibilityRouter returns a chi router for /api/compatibility.
func CompatibilityRouter() http.Handler {
	r := chi.NewRouter()
	r.Post("/", checkCompatibility)
	return r
}

// POST /api/compatibility
func checkCompatibility(w http.ResponseWriter, r *http.Request) {
	var req compatibilityRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || req.Parts == nil {
		writeError(w, http.StatusBadRequest, `"parts" object is required`)
		return
	}

	issues := make([]string, 0)
	parts := req.Parts

	cpu := parts["CPU"]
	gpu := parts["GPU"]
	psu := parts["PSU"]
	caseComp := parts["Case"]
	mb := parts["Motherboard"]

	// CPU ↔ Motherboard socket
	if mb.Details != nil && cpu.Details != nil {
		mbSocket := detailStr(mb, "Socket")
		cpuSocket := detailStr(cpu, "Socket")
		if mbSocket != "" && cpuSocket != "" && mbSocket != cpuSocket {
			issues = append(issues, fmt.Sprintf(
				"Motherboard not compatible with CPU: socket %s ≠ %s", mbSocket, cpuSocket,
			))
		}
	}

	// GPU card length ↔ Case max GPU length
	if gpu.Details != nil && caseComp.Details != nil {
		gpuLen := detailInt(gpu, "Card Length")
		caseMax := detailInt(caseComp, "Max GPU Length")
		if gpuLen > 0 && caseMax > 0 && gpuLen > caseMax {
			issues = append(issues, fmt.Sprintf(
				"GPU not compatible with Case: card length %dmm exceeds case max %dmm", gpuLen, caseMax,
			))
		}
	}

	// Motherboard form factor ↔ Case motherboard support
	if mb.Details != nil && caseComp.Details != nil {
		mbFF := detailStr(mb, "Form Factor")
		caseMbSupport := detailStr(caseComp, "Motherboard Support")
		if mbFF != "" && caseMbSupport != "" {
			supported := strings.Split(caseMbSupport, "/")
			for i, s := range supported {
				supported[i] = strings.TrimSpace(s)
			}
			if !containsStr(supported, mbFF) {
				issues = append(issues, fmt.Sprintf(
					"Motherboard not compatible with Case: %s not supported (supports %s)", mbFF, caseMbSupport,
				))
			}
		}
	}

	// PSU wattage ↔ CPU + GPU TDP
	if psu.Details != nil && (cpu.Details != nil || gpu.Details != nil) {
		psuW := detailInt(psu, "Wattage")
		var cpuTdp, gpuTdp int
		if cpu.Details != nil {
			cpuTdp = detailInt(cpu, "TDP")
		}
		if gpu.Details != nil {
			gpuTdp = detailInt(gpu, "TDP")
		}
		totalW := cpuTdp + gpuTdp
		if psuW > 0 && totalW > 0 && psuW < totalW {
			issues = append(issues, fmt.Sprintf(
				"PSU not compatible with System: %dW < CPU (%dW) + GPU (%dW) = %dW",
				psuW, cpuTdp, gpuTdp, totalW,
			))
		}
	}

	writeJSON(w, http.StatusOK, compatibilityResponse{Issues: issues})
}

// detailStr returns a string representation of a component detail field.
func detailStr(p partDetails, key string) string {
	if p.Details == nil {
		return ""
	}
	v, ok := p.Details[key]
	if !ok {
		return ""
	}
	return fmt.Sprintf("%v", v)
}

// detailInt parses the leading integer from a detail field value (handles "65W", "300mm", plain 65).
func detailInt(p partDetails, key string) int {
	s := detailStr(p, key)
	var buf strings.Builder
	for _, c := range s {
		if c >= '0' && c <= '9' {
			buf.WriteRune(c)
		} else if buf.Len() > 0 {
			break
		}
	}
	n, _ := strconv.Atoi(buf.String())
	return n
}

func containsStr(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}

File: backend/routes/compatibility.js
const { Router } = require('express');

const router = Router();

// POST /api/compatibility
// Body: { parts: { CPU: { details: {...} }, GPU: { details: {...} }, ... } }
router.post('/', (req, res) => {
  const { parts } = req.body;
  if (!parts || typeof parts !== 'object') {
    return res.status(400).json({ error: '"parts" object is required' });
  }

  const issues = [];
  const { CPU: cpu, GPU: gpu, PSU: psu, Case: caseComp, Motherboard: mb } = parts;

  if (mb && cpu) {
    const mbSocket = String(mb.details?.['Socket'] ?? '');
    const cpuSocket = String(cpu.details?.['Socket'] ?? '');
    if (mbSocket && cpuSocket && mbSocket !== cpuSocket) {
      issues.push(`Motherboard not compatible with CPU: socket ${mbSocket} ≠ ${cpuSocket}`);
    }
  }

  if (gpu && caseComp) {
    const gpuLen = parseInt(String(gpu.details?.['Card Length'] ?? '0'));
    const caseMax = parseInt(String(caseComp.details?.['Max GPU Length'] ?? '0'));
    if (gpuLen > 0 && caseMax > 0 && gpuLen > caseMax) {
      issues.push(`GPU not compatible with Case: card length ${gpuLen}mm exceeds case max ${caseMax}mm`);
    }
  }

  if (mb && caseComp) {
    const mbFF = String(mb.details?.['Form Factor'] ?? '');
    const caseMbSupport = String(caseComp.details?.['Motherboard Support'] ?? '');
    const supported = caseMbSupport.split('/').map((s) => s.trim());
    if (mbFF && supported.length > 0 && !supported.includes(mbFF)) {
      issues.push(`Motherboard not compatible with Case: ${mbFF} not supported (supports ${caseMbSupport})`);
    }
  }

  if (psu && (cpu || gpu)) {
    const psuW = parseInt(String(psu.details?.['Wattage'] ?? '0'));
    const cpuTdp = cpu ? parseInt(String(cpu.details?.['TDP'] ?? '0')) : 0;
    const gpuTdp = gpu ? parseInt(String(gpu.details?.['TDP'] ?? '0')) : 0;
    const totalW = cpuTdp + gpuTdp;
    if (psuW > 0 && totalW > 0 && psuW < totalW) {
      issues.push(`PSU not compatible with System: ${psuW}W < CPU (${cpuTdp}W) + GPU (${gpuTdp}W) = ${totalW}W`);
    }
  }

  res.json({ issues });
});

module.exports = router;
